### PR TITLE
[SPARK-45481][SQL] Introduce a mapper for parquet compression codecs

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
Currently, Spark supported all the parquet compression codecs, but the parquet supported compression codecs and spark supported are not completely one-on-one due to Spark introduce a fake compression codecs none.
On the other hand, there are a lot of magic strings copy from parquet compression codecs. This issue lead to developers need to manually maintain its consistency. It is easy to make mistakes and reduce development efficiency.

The `CompressionCodecName`, refer: https://github.com/apache/parquet-mr/blob/master/parquet-common/src/main/java/org/apache/parquet/hadoop/metadata/CompressionCodecName.java

### Why are the changes needed?
Let developers easy to use parquet compression codecs.

### Does this PR introduce _any_ user-facing change?
'No'.
Introduce a new class.

### How was this patch tested?
Exists test cases.

### Was this patch authored or co-authored using generative AI tooling?
'No'.

### CODE REVIEW

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetCompressionCodecMapper.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  UNCOMPRESSED(CompressionCodecName.UNCOMPRESSED),
+  SNAPPY(CompressionCodecName.SNAPPY),
+  GZIP(CompressionCodecName.GZIP),
+  LZO(CompressionCodecName.LZO),
+  BROTLI(CompressionCodecName.BROTLI),
+  LZ4(CompressionCodecName.LZ4),
+  ZSTD(CompressionCodecName.ZSTD);
+
+  // There are some parquet supported compression codec that doesn't supported by Spark.
+  // LZ4_RAW(CompressionCodecName.LZ4_RAW)
```
Comment (LuciferYang) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetCompressionCodecMapper.java:38]: https://github.com/apache/spark/pull/41507

IIRC, LZ4_RAW is already supported？

Comment (beliefer) [reply sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetCompressionCodecMapper.java:38]: Thank you for the reminder. I will check it.

**Comment by srowen** [comment]
Can we retrigger tests?

**Comment by beliefer** [comment]
ping @dongjoon-hyun @srowen @wangyum

**Comment by beliefer** [comment]
The GA failure is unrelated.

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetCompressionCodecMapper.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+import java.util.Locale;
+
+import org.apache.parquet.hadoop.metadata.CompressionCodecName;
+
+/**
+ * A mapper class easy to obtain compression codecs based on their names.
+ */
+public enum ParquetCompressionCodecMapper {
+  NONE(null),
+  UNCOMPRESSED(CompressionCodecName.UNCOMPRESSED),
```
Comment (srowen) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetCompressionCodecMapper.java:29]: Why make our own enum if there is already an enum-like list of codecs in parquet?

Comment (beliefer) [reply sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetCompressionCodecMapper.java:29]: One reason is Spark add the compression codecs `none`, another is out-of-date. Before https://github.com/apache/spark/pull/43310, the parquet supported compression codecs and spark supported are not completely one-on-one.

Comment (srowen) [reply sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetCompressionCodecMapper.java:29]: Ok, looks fine

**Comment by beliefer** [comment]
cc @dongjoon-hyun

**Comment by beliefer** [comment]
cc @viirya

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetIOSuite.scala [hunk 3]
```diff
@@ -857,12 +855,12 @@ class ParquetIOSuite extends QueryTest with ParquetTest with SharedSparkSession
...
     // Checks default compression codec
     checkCompressionCodec(
-      CompressionCodecName.fromConf(spark.conf.get(SQLConf.PARQUET_COMPRESSION)))
+      ParquetCompressionCodec.fromString(spark.conf.get(SQLConf.PARQUET_COMPRESSION)))
 
-    checkCompressionCodec(CompressionCodecName.UNCOMPRESSED)
-    checkCompressionCodec(CompressionCodecName.GZIP)
-    checkCompressionCodec(CompressionCodecName.SNAPPY)
-    checkCompressionCodec(CompressionCodecName.ZSTD)
+    checkCompressionCodec(ParquetCompressionCodec.UNCOMPRESSED)
```
Comment (LuciferYang) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetIOSuite.scala:860]: Unrelated to this pr, but why were only four types of Compression Codec tested here? Was the test case not modified when a new type was added?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala [hunk 4]
```diff
@@ -2709,31 +2709,32 @@ class HiveDDLSuite
...
-    test(s"SPARK-22158 convertMetastore should not ignore table property - $fileFormat") {
-      withSQLConf(CONVERT_METASTORE_ORC.key -> "true", CONVERT_METASTORE_PARQUET.key -> "true") {
-        withTable("t") {
-          withTempPath { path =>
-            sql(
-              s"""
-                |CREATE TABLE t(id int) USING hive
-                |OPTIONS(fileFormat '$fileFormat', compression '$compression')
-                |LOCATION '${path.toURI}'
+  Seq(("orc", "ZLIB"), ("parquet", ParquetCompressionCodec.GZIP.name)).foreach {
```
Comment (LuciferYang) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala:2712]: nit: Make `Seq(("orc", "ZLIB"), ("parquet", ParquetCompressionCodec.GZIP.name))` a variable, and then use `seq..foreach { case (fileFormat, compression) =>`. Would the code below need to be reformatted?

**Review by LuciferYang — APPROVED** [review]
+1, LGTM

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetIOSuite.scala (hunk 3 above)
Comment (beliefer) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetIOSuite.scala:860]: I tested the other compression codec, the tests failed!
It seems not supported the others yet.

Comment (LuciferYang) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetIOSuite.scala:860]: ok

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala (hunk 4 above)
Comment (beliefer) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala:2712]: For reduce the change here, let's use
```
Seq(
  ("orc", "ZLIB"),
  ("parquet", ParquetCompressionCodec.GZIP.name)).foreach { case (fileFormat, compression) =>
```

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetIOSuite.scala (hunk 3 above)
Comment (beliefer) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetIOSuite.scala:860]: If I have time, I will try to cover these tests.

Comment (beliefer) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetIOSuite.scala:860]: I got it now. `lzo` is supported by cloudera Hadoop. Spark doesn't have it built-in.

**Comment by beliefer** [comment]
The GA failure is unrelated.
Merged to master
@srowen @LuciferYang Thank you!

**Pull request closed without merging**